### PR TITLE
Minecraft new project wizard

### PR DESCRIPTION
**Pull request by Earthcomputer**
This is a complete rewrite of all the project creators in MinecraftDev.

image

Although the rewrite isn't quite 100% complete, it is nearly there, any future changes are likely to be on the level of small tweaks so now is a good time to start reviewing so fixes can go in with those tweaks.

## Notable changes
The following notable changes have been made with this rewrite (aside from like, the entire thing):
- Removed creator support for anything older than Minecraft 1.16.5.
   - ForgeGradle 2 support has been removed.
   - Support for quirks that only affect old Fabric (<1.16.5) versions has been removed.
   - LiteLoader support has been removed from MinecraftDev entirely. The newest version of LiteLoader is for 1.12.2.
   - Sponge 7 (and below) support has been removed.
- Removed dependency on TemplateMakerFabric in favour of an in-house solution (it was only used to download Fabric versions).
- Certain more exotic features of the old Fabric creator have been removed from the UI and hardcoded. They were hard to use in comparison to simply modifying the code after the project has been created.

## Extension points
This PR adds the following extension points:
- `platformTypeWizard` for adding additional platform types. Built-in platform types are "Mod" and "Plugin".
- `modPlatformWizard` for adding additional mod creators. Built-in mod creators are "Forge", "Fabric" and "Architectury".
- `pluginPlatformWizard` for adding additional plugin creators. Built-in plugin creators are "Sponge", "Bukkit", "Velocity" and "BungeeCord".
- `bukkitPlatformWizard` for adding additional Bukkit creators to appear under the Bukkit submenu. Built-in Bukkit creators are "Spigot" and "Paper".
- `bungeePlatformWizard` for adding additional BungeeCord creators to appear under the BungeeCord submenu. Built-in BungeeCord creators are "BungeeCord" and "Waterfall".
- `projectSetupFinalizer` to run additional steps after the project has been setup.
- `buildSystemWizard` to add new build systems to project creators. Built-in build systems are "Gradle" and "Maven".
- `buildSystemSupport` implements a specific build system on a specific platform. For example, there is an implementation for both Maven and Gradle on Bukkit, and for Gradle on Forge.

## Steps to completion
- [x] Ensure the correct Java version is used in different situations (e.g. Java 8 on Minecraft 1.16.5 and Java 17 on Minecraft 1.19).
- [x] Ensure the correct Gradle version is used in different situations.
- [x] Ensure run configs are created.
- [x] Generate appropriate gitignores if git is enabled (there should be a checkbox for that but it seems broken).
- [x] Minor UI tweaks (text field width, spacing, alignment)
- [x] Documentation in key areas of the creator API

### CODE REVIEW

File: src/main/kotlin/creator/ProjectSetupFinalizerWizardStep.kt [hunk 1]
```diff
@@ -110,162 +85,57 @@ interface ProjectSetupFinalizer {
...
     }
 
-    override fun validateConfigs(creator: MinecraftProjectCreator, context: WizardContext): Boolean {
-        config = creator.config
-
-        if (creator.buildSystem !is GradleBuildSystem) {
-            return true
+    override fun setupUI(builder: Panel) {
+        with(builder) {
+            row("JDK Version:") {
```
Comment (Mysterious-Dev) [inline src/main/kotlin/creator/ProjectSetupFinalizerWizardStep.kt]: Maybe change `JDK Version :` to `JDK :` to fit with other types of project wizard.

image
image

**Review by Mysterious-Dev — APPROVED** [review]
Great job ! I just have a suggestion for a detail.

**Comment by Earthcomputer** [comment]
Unfortunately I had to rewrite the way version selection works in order to fix widgets not aligning with each other. But the UI is looking good now.

**Comment by Earthcomputer** [comment]
This PR is now done :+1:

**Review by Mysterious-Dev — APPROVED** [review]
LGTM ^^. Great job 👍.

**Comment by Mysterious-Dev** [comment]
Maybe update `gradle-intellij-plugin` to v1.13.0 ?

https://github.com/JetBrains/gradle-intellij-plugin/releases/tag/v1.13.0

**Comment by Earthcomputer** [comment]
That is out of scope for this PR

**Comment by Mysterious-Dev** [comment]
OK 👍

**Review by RedNesto — APPROVED** [review]
Code looks fine 👍
We will need to remove Sponge 7 from the available versions when the new wizard lands in a release (having it broken in nightly is fine for me.)

**Comment by Earthcomputer** [comment]
That's easily doable, I'm already doing it for Forge, filtering out old MC versions

**Comment by T3sT3ro** [comment]
Does it validate/slugify/otherwise handle generation of packages, class etc. from names in inappropriate format (for example properties with hyphens, apostrophes, quotes etc.)? Asking, because current wizards can generate invalid files (when artifactID is a `hyphenated-package-format`, when mod's name contains single/double quotes, then invalid packages, classes and symbols generate). I guess most of the responsibility lies on specific builders.

**Comment by Earthcomputer** [comment]
It should be better at normalizing invalid names, yes.

**Comment by DenWav** [comment]
This looks really great. When I test it though the UI is very slow and laggy - is this just some weirdness I'm experiencing, or does that happen for anyone else too?

Okay tested a bit more and it seems fine. Typing in the project name is slow, probably due to the validation I guess.

**Review by DenWav — APPROVED** [review]

**Comment by Earthcomputer** [comment]
Huh that's strange, I get no performance issues typing the project name. I guess we can see if it's something people have problems with, but it's not the end of the world. While testing it I found a bug which I just fixed.

**Comment by T3sT3ro** [comment]
On my PC it is only slight stuttering, sometimes more obvious, sometimes less. I press quickly `"ASD ` and it's mostly evident in this case.

After performing "a monkey test" on the name field I noticed, that "Main Class" is not properly cleared.

I noticed, that in this case, when:
1. Name has been cleared
2. Main Class field got filled with garbage
3. I cleared Main Class field
4. I pressed backspace on now empty Name field

The Main Class field got populated with garbage again.

Maybe add a debounce for validation to mitigate stuttering and generate fields with a little delay? Say, 0.2s

**Comment by Earthcomputer** [comment]
If your project name contained periods ('.') it would break the main class field until the last commit. Were you testing on the last commit?

**Comment by T3sT3ro** [comment]
~~yup, I'm on `83949cce`~~ scratch that, I'm not, I'll test once more in a sec

**@EDIT**
Now I'm on `83949cce`.
the field no longer breaks for monkey test.

There is still some lag input, but I feel like it occurs from time to time, and even less with passing time. Mostly evident in the beginning when the window is freshly opened. It's hard to reproduce, I have quite decent hardware, it might be more noticeable on older machines.

For me, the lag input is mostly evident when I quickly type and then quickly erase with backspace.

**Comment by DenWav** [comment]
Unfortunately the "window recently opened" thing is common for IntelliJ, due to how it loads UI elements on-demand (and it often does a lot more work than necessary). Honestly not sure how much of that is fixable.

**Comment by Earthcomputer** [comment]
I believe it's something that JetBrains is improving over time.

**Pull request merged**